mm2src/lp_ordermatch/simple_market_maker.rs

```diff
@@ -0,0 +1,702 @@
+            StartSimpleMakerBotError::Transport(_) | StartSimpleMakerBotError::InternalError(_) => {
+                StatusCode::INTERNAL_SERVER_ERROR
+            },
+        }
+    }
+}
+
+impl HttpStatusCode for StopSimpleMakerBotError {
+    fn status_code(&self) -> StatusCode {
+        match self {
+            // maybe bad request is not adapted for the first errors.
+            StopSimpleMakerBotError::AlreadyStopped | StopSimpleMakerBotError::AlreadyStopping => {
+                StatusCode::BAD_REQUEST
+            },
+            StopSimpleMakerBotError::Transport(_) | StopSimpleMakerBotError::InternalError(_) => {
+                StatusCode::INTERNAL_SERVER_ERROR
+            },
+        }
+    }
+}
+
+struct TradingPair {
+    base: String,
+    rel: String,
+}
+
+impl TradingPair {
+    pub fn new(base: String, rel: String) -> TradingPair { TradingPair { base, rel } }
+
+    pub fn as_combination(&self) -> String { self.base.clone() + "/" + self.rel.clone().as_str() }
+}
+
+pub async fn tear_down_bot(ctx: MmArc) {
+    let simple_market_maker_bot_ctx = TradingBotContext::from_ctx(&ctx).unwrap();
+    {
+        let mut trading_bot_cfg = simple_market_maker_bot_ctx.trading_bot_cfg.lock().await;
+        trading_bot_cfg.clear();
+    }
+    cancel_pending_orders(&ctx).await;
+}
+
+async fn get_non_zero_balance(coin: MmCoinEnum) -> Result<MmNumber, MmError<OrderProcessingError>> {
+    let coin_balance = match coin.my_balance().compat().await {
+        Ok(coin_balance) => coin_balance,
+        Err(err) => {
+            warn!("err with balance: {} - reason: {}", coin.ticker(), err.to_string());
+            return MmError::err(OrderProcessingError::BalanceInternalError);
+        },
+    };
+    if coin_balance.spendable.is_zero() {
+        warn!("balance for: {} is zero", coin.ticker());
+        return MmError::err(OrderProcessingError::BalanceIsZero);
+    }
+    Ok(MmNumber::from(coin_balance.spendable))
+}
+
+async fn vwap_calculation(
+    kind: VwapCalculationSide,
+    swaps_answer: MyRecentSwapsAnswer,
+    nb_valid_trades: &mut usize,
+    cfg: &SimpleCoinMarketMakerCfg,
+    calculated_price: MmNumber,
+) -> MmNumber {
+    let mut average_trading_price = calculated_price.clone();
+    let mut total_sum_price_volume = MmNumber::default();
+    let mut total_volume = MmNumber::default();
+    for swap in swaps_answer.swaps.iter() {
+        if !swap.is_finished_and_success() {
+            *nb_valid_trades -= 1;
+            continue;
+        }
+        let (my_amount, other_amount) = match swap.get_my_info() {
+            Some(x) => (MmNumber::from(x.my_amount), MmNumber::from(x.other_amount)),
+            None => {
+                *nb_valid_trades -= 1;
+                continue;
+            },
+        };
+        // todo: refactor to a function
+        let cur_sum_price_volume = match kind {
+            VwapCalculationSide::Base => {
+                let cur_price = my_amount / other_amount.clone();
```

Comment:
To avoid excessive cloning please consider using borrow `&my_amount / &other_amount`